libs/contrib/Data/String/Parser.idr

```diff
@@ -290,3 +291,75 @@ integer : Monad m => ParseT m Integer
 integer = do minus <- succeeds (char '-')
              x <- some digit
              pure $ if minus then (intFromDigits x)*(-1) else intFromDigits x
+
+
+||| Parse repeated instances of at least one `p`, separated by `s`,
+||| returning a list of successes.
+|||
+||| @ p the parser for items
+||| @ s the parser for separators
+export
+covering
+sepBy1 : Monad m => (p : ParseT m a)
+                 -> (s : ParseT m b)
+                 -> ParseT m (List a)
+sepBy1 p s = [| p :: many (s *> p) |]
+
+||| Parse zero or more `p`s, separated by `s`s, returning a list of
+||| successes.
+|||
+||| @ p the parser for items
+||| @ s the parser for separators
+export
+covering
+sepBy : Monad m => (p : ParseT m a)
+                -> (s : ParseT m b)
+                -> ParseT m (List a)
+sepBy p s = (p `sepBy1` s) <|> pure []
+
+||| Parses /one/ or more occurrences of `p` separated by `comma`.
+export
+covering
+commaSep1 : Monad m => ParseT m a -> ParseT m (List a)
+commaSep1 p = p `sepBy1` (char ',')
+
+||| Parses /zero/ or more occurrences of `p` separated by `comma`.
+export
+covering
+commaSep : Monad m => ParseT m a -> ParseT m (List a)
+commaSep p = p `sepBy` (char ',')
+
+||| Run the specified parser precisely `n` times, returning a vector
+||| of successes.
+export
+ntimes : Monad m => (n : Nat) -> ParseT m a -> ParseT m (Vect n a)
+ntimes    Z  p = pure Vect.Nil
+ntimes (S n) p = [| p :: (ntimes n p) |]
+
+||| Parses a letter or digit (a character between \'0\' and \'9\').
+||| Returns the parsed character.
+export
+alphaNum : Monad m => ParseT m Char
+alphaNum = satisfy isAlphaNum <?> "letter or digit"
+
+||| Parses a letter (an upper case or lower case character). Returns the
+||| parsed character.
+export
+letter : Monad m => ParseT m Char
+letter = satisfy isAlpha <?> "letter"
+
+||| Succeeds if and only if the argument parser fails.
+|||
+||| In Parsec, this combinator is called `notFollowedBy`.
+export
+requireFailure : Monad m => (p : ParseT m a)
```

Comment:
```suggestion
requireFailure : Functor m => ParseT m a -> ParseT m ()
requireFailure (P runParser) = P $ \s => reverseResult s <$> runParser s
where
  reverseResult : State -> Result a -> Result ()
  reverseResult  s             (Fail _ _) = OK () s
  reverseResult (S _ _ maxPos) (OK _ _)   = Fail maxPos "Purposefully changed OK to Fail" -- not sure about maxPos
```

Reply:
Um, dunno how to make a multiline suggestion..

Reply:
This version looks much better! Updated.